File: 프로그래머스_모의고사/2차_문제집/송기훈/3번.py
# 18.8점
# import sys

# sys.setrecursionlimit(10**9)

# def solution(n, roads, sources, destination):
#     answer = []
#     matrix = [[] for _ in range(n + 1)]

#     for r in roads:
#         a = r[0]
#         b = r[1]
#         matrix[r[0]].append(b)
#         matrix[r[1]].append(a)

#     def dfs(source, depth):
#         if destination in matrix[source]:
#             return depth
#         for i in matrix[source]:
#             return dfs(i, depth + 1)
#         return -1

#     for s in sources:
#         if s == destination:
#             answer.append(0)
#         else:
#             answer.append(dfs(s, 1))

#     return answer

from collections import deque


def solution(n, roads, sources, destination):
    answer = []
    matrix = [[] for _ in range(n + 1)]

    for r in roads:
        a = r[0]
        b = r[1]
        matrix[a].append(b)
        matrix[b].append(a)

    q = deque()
    visited = [-1 for _ in range(n + 1)]
    visited[destination] = 0
    q.append((destination, 0))
    while q:
        now, depth = q.popleft()
        for n in matrix[now]:
            if visited[n] < 0:
                q.append((n, depth + 1))
                visited[n] = depth + 1

    for s in sources:
        answer.append(visited[s])

    return answer

File: 프로그래머스_모의고사/2차_문제집/이아름/3번.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination) {
	vector<int> answer;
	vector<vector<int>> connect(n + 1);
	for (auto r : roads) {
		connect[r[0]].push_back(r[1]);
		connect[r[1]].push_back(r[0]);
	}
	vector<int> distance(n + 1, -1);
	queue<vector<int>> q; //위치, 거리
	q.push({ destination,0 });
	distance[destination] = 0;
	while (!q.empty()) {
		vector<int> now = q.front(); q.pop();
		for (int n : connect[now[0]]) {
			if (distance[n] < 0) {
				q.push({ n,now[1] + 1 });
				distance[n] = now[1] + 1;
			}
		}
	}
	for (int s : sources) {
		answer.push_back(distance[s]);
	}

	return answer;
}
